### ISSUE TITLE
Allow JavaScript code to appear at the top level of an extending template

### ISSUE THREAD

**Issue by KSXGitHub** [pugjs/pug#2764]
It would be cool if we are able to do this:

*index.pug*
```pug
extends layout.pug
- const message = 'Hello, World!!'
block title
  | !{message}
block body
  h1= message.toUpperCase()
```

*layout.pug*
```pug
doctype html
html
  head
    title
      block title
  body
    block body
```

*Expected result: index.html*
```html
<!DOCTYPE html>
<html>
  <head>
    <title>Hello, World!!</title>
  </head>
  <body>
    <h1>HELLO, WORLD!!</h1>
  </body>
</html>
```

If mixins can appear at the top level of an extending template, why can't JavaScript code?

**Comment by alexbosquin** [comment]
Hi bro, i think to use javascript you need to use with this syntax:
`
script.
      const message = "hello world";
`

if you have any trouble use this website: http://html2jade.org/

**Comment by KSXGitHub** [comment]
@alexbosquin No, I'm not talking about embedding html `<script>` tag that would later run in browser, I'm talking about JavaScript code that executes during compile time, in Pug, in Node.js.

I expect Pug to generate actual text in the output `index.html`, not the `<script>` tag

**Comment by ForbesLindesay** [comment]
It might be neat to support something like this.  The main use case generally seems to be declaring variables.  The question is: how should those variables be scoped, and when should we actually declare them?  i.e. what happens if layout.pug also declares a message variable?

**Comment by KSXGitHub** [comment]
@ForbesLindesay 

> what happens if layout.pug also declares a message variable?

My expectation is: The one that is declared in `index.pug` is local, the one that is declared in `layout.pug` is global, therefore, `message` in `index.pug` refers to the local one. BTW, how exactly do mixins behave in this case?

In case you're asking how to implement it, frankly, I didn't contribute to this project before so I don't know how exactly to implement it. My idea is to wrap all extending blocks in `index.pug` into a function which in turn exports an object contains code for the blocks. Here's my hypothetical code to make it clear (it's "hypothetical", not even JavaScript, this is but a suggestion):

```javascript
// from index.pug

// above all
const _blocks = (() => {
  const result = {}

  // from: - const message = 'Hello, World!!'
  const message = 'Hello, World!!'

  // from: block title
  result['title'] = message

  // from: block body
  result['body'] = `<h1>${message.toUpperCase()}</h1>`

  return result
})()

// main
return `
<html>
  <head>
    <title>
      ${_blocks['title']}
    </title>
  </head>
  <body>
    ${_blocks['body']}
  </body>
</html>
`
```

**Comment by ForbesLindesay** [comment]
Unfortunately, the blocks are completely compiled away, so we'd actually have to implement that by going through all of the index.pug file and renaming references to `message` to something like `pug_local_variable_index_message`.  It is an option.  The downside of doing it that way is that another thing people often want a "better" way to solve is:

index.pug
```pug
extends layout.pug

block variables
  - const title = "index";
```

layout.pug
```pug
block variables
  - const title = 'default title';

html
  head
    title= title
```

**Comment by KSXGitHub** [comment]
@ForbesLindesay 

> so we'd actually have to implement that by going through all of the index.pug file and renaming references to `message` to something like `pug_local_variable_index_message`

Seriously, don't do this, this might lead to bugs and ton of unwanted behaviour that users have to take care of.

The code you shown is, of course, great idea, except users shouldn't have to write `block variables` themselves but pug should add them before compile:

**Original files**

File: index.pug
```pug
extends layout.pug
- const a = 'a'
block display-a
  p= a
- const b = 'b'
block display-b
  p= b
```

File: layout.pug
```pug
html
  body
    block display-a
    block display-b
```

**Load files and add "virtual" blocks**


Buffer: index.pug
```pug
extends layout.pug
block _javascript_block_0
  - const a = 'a' 
block display-a
  p= a
block _javascript_block_1
  - const b = 'b'
block display-b
  p= b
```

Buffer: layout.pug
```pug
//- Added by pug
block _javascript_block_0
block _javascript_block_1

//- Loaded from layout.pug
html
  body
    block display-a
    block display-b
```

**Comment by ForbesLindesay** [comment]
The renaming is the only way to do scoping.  We can do it very reliably, so users won't see any broken behaviour.  It will have a performance cost at compile time.

**Comment by KSXGitHub** [comment]
@ForbesLindesay Why is renaming the only way? Adding blocks is a nice hack, how is that not possible?

**Comment by KSXGitHub** [comment]
Oh, my bad, I forgot that blocks and mixins don't create their own JavaScript scope. May be renaming variable is the only possible way to do scoping :smile:

**Comment by ForbesLindesay** [comment]
It's not that blocks & mixins don't create their own scope, it's that there isn't any one scope that's shared by all the blocks in a given file, and not by the parent/layout file.
